### PR TITLE
Cleanup array

### PR DESCRIPTION
**Pull request by fsaintjacques**
I'd like one or more pair of eyes to verify this set of changes.

### CODE REVIEW

**Comment by lemire** [comment]
Will review. Soon. Not now.

File: include/util.h [hunk 1]
```diff
@@ -1,58 +1,66 @@
-
 #ifndef UTIL_H
- 
 #define UTIL_H
 
+#define clamp(value, min, max) \
```
Comment (fsaintjacques) [inline include/util.h]: These 3 macros are not side-effect free, I'll need to transform them in static inline functions.

**Comment by lemire** [comment]
Sorry about the delay. I have been dealing with other non-software issues. ;-)

File: include/util.h [hunk 2]
```diff
@@ -1,58 +1,66 @@
...
+    ((value < min) ? min : (value > max) ? max : value)
 
-int32_t binarySearch(uint16_t* source, int32_t n, uint16_t target);
-
+#define roar_min(a, b) (a < b) ? a : b;
+#define roar_max(a, b) (a < b) ? b : a;
 
-int32_t advanceUntil(uint16_t * array, int32_t pos, int32_t length, uint16_t min);
```
Comment (lemire) [inline include/util.h]: Might be useful to document that ```advanceUntil``` is possibly implemented as an exponential search.

File: src/containers/array.c [hunk 3]
```diff
@@ -12,799 +12,357 @@
...
+    } else {
+        free(array);
+        container->array = malloc(new_capacity * sizeof(uint16_t));
     }
+
     // TODO: handle the case where realloc fails
-    if(arr->array == NULL) {
-    	printf("Well, that's unfortunate. Did I say you could use this code in production?\n");
-    }
+    assert(container->array != NULL);
```
Comment (lemire) [inline src/containers/array.c]: The assert is fine here, but, to be fair, it is not what one would want in production (not that we should worry about that here).

**Comment by lemire** [comment]
This all looks fine to me. I recommend merging.

File: include/util.h (hunk 2 above)
Comment (fsaintjacques) [inline include/util.h]: I didn't touch advanceUntil, this is an artefact from auto-formating.

Comment (lemire) [inline include/util.h]: I was aware of that, I was just making a general comment as much for myself as for others.

**Pull request merged**